cpp/src/arrow/table_test.cc

```diff
@@ -406,6 +407,163 @@ TEST_F(TestTable, ConcatenateTables) {
   ASSERT_RAISES(Invalid, ConcatenateTables({t1, t3}, &result));
 }
 
+using TestPromoteTableToSchema = TestTable;
+
+TEST_F(TestPromoteTableToSchema, IdenticalSchema) {
+  const int length = 10;
+  MakeExample1(length);
+  std::shared_ptr<Table> table = Table::Make(schema_, arrays_);
+
+  std::shared_ptr<Table> result;
+  auto metadata =
+      std::shared_ptr<KeyValueMetadata>(new KeyValueMetadata({"foo"}, {"bar"}));
```

Comment:
Please avoid `new`
```suggestion
      std::make_shared<KeyValueMetadata>(KeyValueMetadata({"foo"}, {"bar"}));
```

Reply:
This doesn't work. std::forward<> doesn't know what type to forward. (f https://godbolt.org/z/AlKvrO